### PR TITLE
Board page created

### LINKED ISSUE #74: Fullstack: BoardPage
## User story

**Who:** **As a** User
**What:** **I want to** be able see the Board Page 
**Why:** **so that I can**  see the candidates that decide as an owner and I should also be able rank the candidates in a different order if I am an owner or a member. I will also want to be able to check the results by clicking on the Result button and accessing the Menu on the header of the page and see the options according to my user(owner or member). And finally to access the Members Modal View where all the members should be listed.

## Acceptance criteria

The page should look like this if the user is the Owner of the Board
Screenshot 2021-07-30 at 13 53 02

The page should look like this if the user is the Member of the Board
Screenshot 2021-07-30 at 13 52 22

The page should look like this either for the Owner or Member of the Board when the deadline is past due.
Screenshot 2021-07-30 at 13 53 10

When clicking on the Members button we should be able to call the Modal view and list all the members like shown below
Screenshot 2021-07-30 at 14 13 48Screenshot 2021-07-30 at 14 13 18

#### -> React Containers and Components 
- [x] Dummy Page - https://github.com/HackYourFuture-CPH/simply-name.it/issues/44
    1. [ ] BoardPage - src/client/containers/BoardPage/BoardPage.Container.js
    2. [ ] BoardPage - src/client/containers/BoardPage/BoardPage.Style.css
- [x] Header - https://github.com/HackYourFuture-CPH/simply-name.it/issues/15
- [x] Images - https://github.com/HackYourFuture-CPH/simply-name.it/pull/72
- [x] Title
- [x] Button - https://github.com/HackYourFuture-CPH/simply-name.it/issues/2
- [x] Input - https://github.com/HackYourFuture-CPH/simply-name.it/issues/9         
- [x] Button -  https://github.com/HackYourFuture-CPH/simply-name.it/issues/2 (probably needs some adaptation)
- [x] Card and List of Cards - https://github.com/HackYourFuture-CPH/simply-name.it/issues/18
- [x] Button -  https://github.com/HackYourFuture-CPH/simply-name.it/issues/2

#### -> Endpoints
- [x] Endpoint to update the ballots table - https://github.com/HackYourFuture-CPH/simply-name.it/issues/57
- [x] Endpoint to retrieve the result of a certain board - https://github.com/HackYourFuture-CPH/simply-name.it/issues/54
- [x] Endpoint to retrieve the list of all the members and owner of a specific board - https://github.com/HackYourFuture-CPH/simply-name.it/issues/75
- [x] Endpoint to get all the candidates from a board - https://github.com/HackYourFuture-CPH/simply-name.it/issues/64

### PR DESCRIPTION
**Pull request by Hani-far**
# Description

Please provide a short summary explaining what this PR is about.

Fixes #74 

# How to test?

Please provide a short summary how your changes can be tested?

# Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have followed the name conventions for CSS Classnames and filenames, Components names and filenames, Style filenames, if you are in doubt check the the project README.MD and here https://github.com/HackYourFuture-CPH/curriculum/blob/master/review/review-checklist.md
- [ ] I have commented my code, particularly in hard-to-understand areas, if you code was simple enough mark the box anyway
- [ ] I have made corresponding changes to the documentation, if you code was simple enough mark the box anyway
- [ ] This PR is ready to be merged and not breaking any other functionality

### CODE REVIEW

File: src/client/containers/BoardPage/BoardPage.container.js
Comment (Hani-far) [reply src/client/containers/BoardPage/BoardPage.container.js:28]: Hi, thanks Niels. We are working on that

Comment (Hani-far) [reply src/client/containers/BoardPage/BoardPage.container.js]: Hi Niels,
We made 2 separate components for the Owner view and the Member view.

File: src/client/containers/BoardPage/MemberBoardPage.js
Comment (senner007) [inline src/client/containers/BoardPage/MemberBoardPage.js:28]: Similar duplication here. Please refactor. This also contains several subcomponents.
I see components for atleast:
Header
Title
Candidate List

**Review by senner007 — CHANGES_REQUESTED** [review]

File: src/client/containers/BoardPage/BoardPage.container.js
Comment (senner007) [inline src/client/containers/BoardPage/BoardPage.container.js:29]: **It is critical that you learn to identify and refactor/eliminate duplication** 
The below code can be rewritten like this

```
return (
   <div>
      {isLoading ? (
        <div>LOADING....</div>
      ) : (
        userId === boardInfo.creatorId && (
          <OwnerBoardPage boardInfo={boardInfo} />
        )
      )}
      {isLoading ? (
        <div>LOADING....</div>
      ) : (
        userId !== boardInfo.creatorId && (
          <MemberBoardPage boardInfo={boardInfo} />
        )
      )}
    </div>
  );
}
```

```
  if (isLoading) {
    return <div>LOADING....</div>
  }

  return (
    userId === boardInfo.creatorId 
      ? <OwnerBoardPage boardInfo={boardInfo}/>
      : <MemberBoardPage boardInfo={boardInfo} />
  )
  
```

**Review by senner007 — CHANGES_REQUESTED** [review]

File: src/client/containers/BoardPage/AddCandidate.js
Comment (senner007) [reply src/client/containers/BoardPage/AddCandidate.js:2]: The board page is a high level container component that contains subcomponents 
https://reactjs.org/docs/thinking-in-react.html

File: src/client/containers/BoardPage/BoardPage.container.js
Comment (Hani-far) [reply src/client/containers/BoardPage/BoardPage.container.js:29]: Done.

File: src/client/containers/BoardPage/AddCandidate.js
Comment (Hani-far) [reply src/client/containers/BoardPage/AddCandidate.js:2]: Regarding the add candidates, we have separate issue #92 Extend the BoardPage and Add Candidates. and I made a separate PR for it as well https://github.com/HackYourFuture-CPH/simply-name.it/pull/169
But as you wanted to have all the changes in this PR, I merged it with this PR. 
I can separate it inside a container component and make a new PR for the existing issue.

Comment (Hani-far) [reply src/client/containers/BoardPage/AddCandidate.js:2]: > Adding candiates and posting them has nothing to do with the board page. It should reside inside a container component that deals with the candidates. Please make a seperate issue and pr for that.

Regarding the add candidates, we have separate issue #92 Extend the BoardPage and Add Candidates. and I made a separate PR for it as well #169
But as you wanted to have all the changes in this PR, I merged it with this PR.
I can separate it inside a container component and make a new PR for the existing issue.

Comment (senner007) [reply src/client/containers/BoardPage/AddCandidate.js:2]: Basically if I look at the structure, I see this:

Member
- BoardHeader
- CandidateList
- ResultButton

Owner
- BoardHeader
- BoardAddCandidate
- CandidateList
- ResultButton

Right?

What goes into these components are seperate prs.

We then need a way for the components to communicate. If the MemberBoard and OwnerBoard is placed in a Provider context then the boards and subsequently the candidates list well be reloaded when a new candidate is added.
So to recap: Inside BoardAddCandidate a post request is made to create a new candidate and then the state inside the provider is toggled, which means its children are reloaded.

BoardProvider
- MemberBoard
- OwnerBoard

Comment (senner007) [reply src/client/containers/BoardPage/AddCandidate.js:2]: Both the BoardAddCandidate and CandidateList are subcomponents that goes into different prs.

**Comment by senner007** [comment]
We need to create the board component inside and authenticated route and a context. 
See the the UserContext and useUser as an example. 
 Just duplicate for now then we will change them later.
BoardContext and useBoard.

Then in app.js we need this structure

```
<AuthenticatedRoute
          exact
          path="/board/:id"
          isAuthenticated={isAuthenticated}
          isLoading={isLoading}
        >
          <BoardContext>
              <Board />
          </BoardContext>
        </AuthenticatedRoute>

```

**Comment by Saidemm** [comment]
> <AuthenticatedRoute
>           exact
>           path="/board/:id"
>           isAuthenticated={isAuthenticated}
>           isLoading={isLoading}
>         >
>           <BoardContext>
>               <Board />
>           </BoardContext>
>         </AuthenticatedRoute>


Thanks Niels, this explanation is very helpful. Yesterday, I made a BoardContext in BoardPage which has a provider function that return :

**Comment by senner007** [comment]
> > 
> 
> Thanks Niels, this explanation is very helpful. Yesterday, I made a BoardContext in BoardPage which has a provider function that return : return userId === boardInfo.creatorId ? (
> <BoardContext.Provider value={{ userInfo, boardInfo }}>
> 
> </BoardContext.Provider>
> ) : (
> <BoardContext.Provider value={{ userInfo, boardInfo }}>
> 
> </BoardContext.Provider>
> ); >
> Do you mean I can mean should I create this boardContext out of BoardPage?



> > 
> 
> Thank you so much Niels, these explanations are very helpful.
> Yesterday, I made a BoardContext in BoardPage which has a provider function that return :
> 
> ```
>          return userId === boardInfo.creatorId ? (
>             <BoardContext.Provider value={{ userInfo, boardInfo }}>
>               <OwnerBoardPage />
>             </BoardContext.Provider>
>           ) : (
>             <BoardContext.Provider value={{ userInfo, boardInfo }}>
>               <MemberBoardPage />
>             </BoardContext.Provider>
>           ); >
> 
>    @senner008 Do you mean I can mean should I create this boardContext out of BoardPage? 
>     Is this <Board /> you've mentioned before the BoardPage we made?
> ```

The above is not necessary I think. 
The above <Board> component contain this:

```
  const { isLoading } = useBoard();
  
  if (isLoading) {
    return <div>LOADING....</div>
  }

  return (
    userId === boardInfo.creatorId 
      ? <OwnerBoardPage boardInfo={boardInfo}/>
      : <MemberBoardPage boardInfo={boardInfo} />
  )
```
  
Inside the BoardProvider we have this:
```
  const [isLoading, setIsLoading] = useState(true)

  return (
    <UserContext.Provider value={{ isLoading, setIsloading }}>
      {children}
    </UserContext.Provider>
  );
}
```

**Comment by senner007** [comment]
Edited above

**Comment by senner007** [comment]
> We need to create the board component inside and authenticated route and a context.
> See the the UserContext and useUser as an example.
> Just duplicate for now then we will change them later.
> BoardContext and useBoard.
> 
> Then in app.js we need this structure
> 
> ```
> <AuthenticatedRoute
>           exact
>           path="/board/:id"
>           isAuthenticated={isAuthenticated}
>           isLoading={isLoading}
>         >
>           <BoardContext>
>               <Board />
>           </BoardContext>
>         </AuthenticatedRoute>
> ```

BoardProvider, not BoardContext as you said

**Comment by Saidemm** [comment]
> > > 
> > 
> > 
> > Thanks Niels, this explanation is very helpful. Yesterday, I made a BoardContext in BoardPage which has a provider function that return : return userId === boardInfo.creatorId ? (
> > <BoardContext.Provider value={{ userInfo, boardInfo }}>
> > </BoardContext.Provider>
> > ) : (
> > <BoardContext.Provider value={{ userInfo, boardInfo }}>
> > </BoardContext.Provider>
> > ); >
> > Do you mean I can mean should I create this boardContext out of BoardPage?
> 
> > > 
> > 
> > 
> > Thank you so much Niels, these explanations are very helpful.
> > Yesterday, I made a BoardContext in BoardPage which has a provider function that return :
> > ```
> >          return userId === boardInfo.creatorId ? (
> >             <BoardContext.Provider value={{ userInfo, boardInfo }}>
> >               <OwnerBoardPage />
> >             </BoardContext.Provider>
> >           ) : (
> >             <BoardContext.Provider value={{ userInfo, boardInfo }}>
> >               <MemberBoardPage />
> >             </BoardContext.Provider>
> >           ); >
> > 
> >    @senner008 Do you mean should I create this boardContext out of BoardPage? 
> >     Is this <Board /> you've mentioned before the BoardPage we made?
> > ```
> 
> The above is not necessary I think.
> The above "Board" component contain this:
> 
> ```
> 
>   return (
>     userId === boardInfo.creatorId 
>       ? <OwnerBoardPage boardInfo={boardInfo}/>
>       : <MemberBoardPage boardInfo={boardInfo} />
>   )
> ```
> 
> Inside the BoardProvider we have this:
> 
> ```
>   const [isLoading, setIsLoading] = useState(true)
> 
>   return (
>     <BoardContext.Provider value={{ isLoading, setIsloading }}>
>       {children}
>     </BoardContext.Provider>
>   );
> }
> ```

Great, I'll try like this.

**Comment by senner007** [comment]
> Great, I'll try like this.

Inside the AddCandidate component we call useBoards and set the loading state when a candidate is posted. This causes the BoardProvider to rerender its children which also causes he CandidateList to reload and do new fetch to get the candidates

**Comment by Saidemm** [comment]
> > Great, I'll try like this.
> 
> Inside the AddCandidate component we call useBoards and set the loading state when a candidate is posted. This causes the BoardProvider to rerender its children which also causes he CandidateList to reload and do new fetch to get the candidates



> > Great, I'll try like this.
> 
> Inside the AddCandidate component we call useBoards and set the loading state when a candidate is posted. This causes the BoardProvider to rerender its children which also causes he CandidateList to reload and do new fetch to get the candidates

If I understand correctly, you want that when we set loading to true (by using the shared context among all children under board) when e.g. a candidate is added, then this will also result in triggering loading the whole board container and it will hide the whole page for a while (instead of just hiding the list of candidates and show loading only there). Wouldn't this be sub-optimal and bad ux?

**Comment by senner007** [comment]
> If I understand correctly, you want that when we set loading to true (by using the shared context among all children under board) when e.g. a candidate is added, then this will also result in triggering loading the whole board container and it will hide the whole page for a while (instead of just hiding the list of candidates and show loading only there). Wouldn't this be sub-optimal and bad ux?

Indeed.
What we will do is just to signal to the candidatelist that it needs to reload. The rest of the page is unaffected. But we still need to show a loader screen when its uploading a new candidate I think.

**Comment by senner007** [comment]
>  But we still need to show a loader screen when its uploading a new candidate I think.

Maybe show a loader in the AddCandidate component and then when its done a loader in the CandidateList component

**Comment by Saidemm** [comment]
> > But we still need to show a loader screen when its uploading a new candidate I think.
> 
> Maybe show a loader in the AddCandidate component and then when its done a loader in the CandidateList component

so for each of these (board, add-candidate and candidate-list) we should use separate loading states to toggel. We can just use the loading state in shared context among all of them?

**Comment by Saidemm** [comment]
> > We need to create the board component inside and authenticated route and a context.
> > See the the UserContext and useUser as an example.
> > Just duplicate for now then we will change them later.
> > BoardContext and useBoard.
> > Then in app.js we need this structure
> > ```
> > <AuthenticatedRoute
> >           exact
> >           path="/board/:id"
> >           isAuthenticated={isAuthenticated}
> >           isLoading={isLoading}
> >         >
> >           <BoardContext>
> >               <Board />
> >           </BoardContext>
> >         </AuthenticatedRoute>
> > ```
> 
> BoardProvider, not BoardContext as you said

When I surround board provider with AuthenticatedRoute, my page remains in loading state. My guess is that I need to sign-in first but I don't know what username and password I should use.

**Comment by senner007** [comment]
> When I surround board provider with AuthenticatedRoute, my page remains in loading state. My guess is that I need to sign-in first but I don't know what username and password I should use.

We might have to fix it. Have you pushed your changes?

**Comment by senner007** [comment]
> so for each of these (board, add-candidate and candidate-list) we should use separate loading states to toggel. We can just use the loading state in shared context among all of them?

Yes, when then CandidateList is done posting it sets the loading state to true, which triggers a reload in CandidateList. The CandidateList listens for changes in the loading state and does a get request when the state changes and th loading is true

**Comment by senner007** [comment]
> > so for each of these (board, add-candidate and candidate-list) we should use separate loading states to toggel. We can just use the loading state in shared context among all of them?
> 
> Yes, when then CandidateList is done posting it sets the loading state to true, which triggers a reload in CandidateList. The CandidateList listens for changes in the loading state and does a get request when the state changes and th loading is true

When the AddCandidate is done, I mean

**Comment by senner007** [comment]
> Not yet, cause it's not working with this so I commented that part out. Maybe I should push them now, then fix the rest later.

Push it with the Authenticated route. Then I will have a look

**Comment by Saidemm** [comment]
> > so for each of these (board, add-candidate and candidate-list), we should use separate loading states to toggle. We can just use the loading state in a shared context among all of them?
> 
> Yes, when then CandidateList is done post, it sets the loading state to true, which triggers a reload in CandidateList. The CandidateList listens for changes in the loading state and does a get request when the state changes and th loading is true

Thank you so much Niels, for your time.

File: src/client/containers/BoardPage/BoardPage.container.js
Comment (senner007) [inline src/client/containers/BoardPage/BoardPage.container.js]: Still using the context state here.
See here:
https://github.com/HackYourFuture-CPH/simply-name.it/pull/233#issuecomment-907803223

File: src/client/containers/BoardPage/BoardPage.container.js
Comment (senner007) [inline src/client/containers/BoardPage/BoardPage.container.js:68]: If the Result button is the same on both members and owner it might as well be placed here to avoid duplication

File: src/client/containers/BoardPage/BoardPage.style.css
Comment (senner007) [inline src/client/containers/BoardPage/BoardPage.style.css]: use lower case kebab case

File: src/client/containers/BoardPage/MemberBoardPage.js
Comment (senner007) [inline src/client/containers/BoardPage/MemberBoardPage.js]: Is there a reson to put all these in divs? The components should contain the necessary styling

File: src/client/containers/BoardPage/MemberBoardPage.js
Comment (senner007) [inline src/client/containers/BoardPage/MemberBoardPage.js]: Candidate List components should not take props. It should call useBoard to get the information

File: src/client/containers/BoardPage/OwnerBoardPage.js
Comment (senner007) [inline src/client/containers/BoardPage/OwnerBoardPage.js:58]: Same here with the divs.

File: src/client/components/ModalViewComponent/membersData.json
Comment (senner007) [inline src/client/components/ModalViewComponent/membersData.json:4]: The data should come from the database instead

File: src/client/components/CandidateCard/CandidateListArray.js
Comment (senner007) [inline src/client/components/CandidateCard/CandidateListArray.js:1]: Data should come from database

**Review by senner007 — CHANGES_REQUESTED** [review]
A few comments :)

File: src/client/containers/BoardPage/ResultButton/ResultButtonPostDeadline.container.js
Comment (senner007) [inline src/client/containers/BoardPage/ResultButton/ResultButtonPostDeadline.container.js]: Its /boards/:boardId/results
Look in app.js

**Review by senner007 — CHANGES_REQUESTED** [review]

File: src/client/containers/BoardPage/ResultButton/ResultButtonPostDeadline.container.js
Comment (senner007) [inline src/client/containers/BoardPage/ResultButton/ResultButtonPostDeadline.container.js:17]: Try it out please and see if it takes you to the result page

**Review by senner007 — CHANGES_REQUESTED** [review]

File: src/client/containers/BoardPage/ResultButton/ResultButtonPostDeadline.container.js
Comment (Saidemm) [reply src/client/containers/BoardPage/ResultButton/ResultButtonPostDeadline.container.js:17]: Sure. I'm waiting for Rasim to resolve his conflicts on his branch and merge otherwise our page is not functioning to test the above

Comment (senner007) [reply src/client/containers/BoardPage/ResultButton/ResultButtonPostDeadline.container.js:17]: I don't understand what you mean? You simply click this above to see if it takes you to the result page. It has nothing to do with anything else

File: src/client/containers/BoardPage/BoardPage.container.js
Comment (Rasim-Deniz) [reply src/client/containers/BoardPage/BoardPage.container.js]: changed it in this pr https://github.com/HackYourFuture-CPH/simply-name.it/pull/254

File: src/client/containers/BoardPage/MemberBoardPage.js
Comment (Rasim-Deniz) [reply src/client/containers/BoardPage/MemberBoardPage.js]: its changed in this pr https://github.com/HackYourFuture-CPH/simply-name.it/pull/254

File: src/client/containers/BoardPage/ResultButton/ResultButtonPostDeadline.container.js
Comment (Hani-far) [reply src/client/containers/BoardPage/ResultButton/ResultButtonPostDeadline.container.js:17]: @senner007 Yes Niels, It goes to the result page
image

File: src/client/components/ModalViewComponent/membersData.json
Comment (Rasim-Deniz) [reply src/client/components/ModalViewComponent/membersData.json:4]: But there won't be any picture.

Comment (senner007) [reply src/client/components/ModalViewComponent/membersData.json:4]: The data should come from the database. Then you can edit your database if you want to send back images

Comment (senner007) [reply src/client/components/ModalViewComponent/membersData.json:4]: Its in the components folder. My fault.

File: src/client/components/CandidateCard/CandidateListArray.js
Comment (senner007) [reply src/client/components/CandidateCard/CandidateListArray.js:1]: Its in the components section. Disregard

**Review by senner007 — APPROVED** [review]
✔
Lets come back to it once you have the other pr merged

**Comment by senner007** [comment]
@Hani-far 
Ill be back to review it in an hour

**Comment by Hani-far** [comment]
> @Hani-far
> Ill be back to review it in an hour

Thank you so much Niels fo all your help 🍀

File: src/client/containers/UseHooks/useCandidates.js
Comment (senner007) [inline src/client/containers/UseHooks/useCandidates.js:31]: Must be thrown in a state changing function

**Comment by senner007** [comment]
@Rasim-Deniz 
Hi. The adding of the candidates doesn't seem to work. Does it work for you? 
If it doesn't have a ballot it does not return the new candidate

**Comment by senner007** [comment]
@Rasim-Deniz 
I can see I get a 401 when I add the candidate

**Comment by Hani-far** [comment]
> @Rasim-Deniz
> Hi. The adding of the candidates doesn't seem to work. Does it work for you?
> If it doesn't have a ballot it does not return the new candidate

@senner007 It's working for me Niels
screen-capture

**Review by senner007 — APPROVED** [review]
> > @Rasim-Deniz
> > Hi. The adding of the candidates doesn't seem to work. Does it work for you?
> > If it doesn't have a ballot it does not return the new candidate
> 
> @senner007 It's working for me Niels
> screen-capture

My fault!
I forgot to change the userId in add candidates.

**Comment by senner007** [comment]
Great job everyone
This has been a lot of work, but you will be rewarded for the experience.
We could clean the code a bit later but essentially all code must be written like this. 
Each component having a single responsibility and all behavior encapsulated within it.

File: src/client/containers/UseHooks/useCandidates.js
Comment (Rasim-Deniz) [reply src/client/containers/UseHooks/useCandidates.js:31]: How can we do it?

Comment (Rasim-Deniz) [reply src/client/containers/UseHooks/useCandidates.js:31]: @senner007 Should I send the response and check outside where I use this useCandidates?

Comment (senner007) [reply src/client/containers/UseHooks/useCandidates.js:31]: No you just have to do it the same way as with the delete method. Throw in a state changing function

Comment (Rasim-Deniz) [reply src/client/containers/UseHooks/useCandidates.js:31]: This useHook's work was returning candidates but not sure how I can send the error also.

Comment (Rasim-Deniz) [reply src/client/containers/UseHooks/useCandidates.js:31]: We were returning response in delete method how is it gonna be here?

**Comment by shpomp** [comment]
go to sleep! 💤 😴 ☁️ ✨ 🌔

**Pull request merged**
